File: connector-framework/src/test/java/org/identityconnectors/common/StringUtilTests.java
/*
 * ====================
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
 *
 * Copyright 2008-2009 Sun Microsystems, Inc. All rights reserved.
 *
 * The contents of this file are subject to the terms of the Common Development
 * and Distribution License("CDDL") (the "License").  You may not use this file
 * except in compliance with the License.
 *
 * You can obtain a copy of the License at
 * http://opensource.org/licenses/cddl1.php
 * See the License for the specific language governing permissions and limitations
 * under the License.
 *
 * When distributing the Covered Code, include this CDDL Header Notice in each file
 * and include the License file at http://opensource.org/licenses/cddl1.php.
 * If applicable, add the following below this CDDL Header, with the fields
 * enclosed by brackets [] replaced by your own identifying information:
 * "Portions Copyrighted [year] [name of copyright owner]"
 * ====================
 */
package org.identityconnectors.common;

import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertFalse;
import static org.testng.Assert.assertTrue;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.Random;

import org.testng.annotations.Test;

public class StringUtilTests {
    // ========================================================================
    // JUnit Tests
    // ========================================================================
    /**
     */
    @Test
    public void testIndexOfDigit() {
        int test = 0;
        final String TEST0 = null;
        final String TEST1 = "fsadlkjffj";
        final String TEST2 = "abac2dafj";
        final String TEST3 = "fa323jf4af";

        test = StringUtil.indexOfDigit(TEST0);
        assertEquals(test, -1);
        test = StringUtil.indexOfDigit(TEST1);
        assertEquals(test, -1);
        test = StringUtil.indexOfDigit(TEST2);
        assertEquals(test, 4);
        test = StringUtil.indexOfDigit(TEST3);
        assertEquals(test, 2);
    }

    /**
     */
    @Test
    public void testIndexOfNonDigit() {
        int test = 0;
        final String TEST0 = null;
        final String TEST1 = "2131398750976";
        final String TEST2 = "21312a9320484";
        final String TEST3 = "32323aa323435";

        test = StringUtil.indexOfNonDigit(TEST0);
        assertEquals(test, -1);
        test = StringUtil.indexOfNonDigit(TEST1);
        assertEquals(test, -1);
        test = StringUtil.indexOfNonDigit(TEST2);
        assertEquals(test, 5);
        test = StringUtil.indexOfNonDigit(TEST3);
        assertEquals(test, 5);
    }

    /**
     */
    @Test
    public void testSubDigitString() {
    }

    /**
     */
    @Test
    public void testStripXmlAttribute() {
        final String DATA[][] = {
                // source, attr, result
                { null, null, null },
                { "attr='fads'", "attr", "" },
                { "at1='fasd' at1=''", "at1", "" }
        };
        String tst = null;
        for (int i = 0; i < DATA.length; i++) {
            tst = StringUtil.stripXmlAttribute(DATA[i][0], DATA[i][1]);
            assertEquals(tst, DATA[i][2]);
        }
    }

    /**
     * Make sure it removes '\n'.
     */
    @Test
    public void testStripNewlines() {
        final String[][] TESTS = new String[][] { { null, null },
                { "afdslf\n", "afdslf" }, { "afds\nfadkfj", "afdsfadkfj" },
                { "afds \nfadkfj", "afds fadkfj" },
                { "afds\n fadkfj", "afds fadkfj" } };
        String tmp;
        for (String[] data : TESTS) {
            tmp = StringUtil.stripNewlines(data[0]);
            assertEquals(tmp, data[1]);
        }
    }

    /**
     */
    @Test
    public void testStripXmlComments() {
        final String DATA[][] = {
                // test data -> result
                { null, null }, { "<!--test1-->", "" },
                { "test data", "test data" },
                { "<!--test data", "<!--test data" },
                { "test data-->", "test data-->" },
                { "test data <!-- fasdkfj -->", "test data " },
                { "<!-- fasdkfj --> test data", " test data" },
                { "<!-- fasdkfj --> test data<!-- fadsom-->", " test data" } };

        String tst = null;
        for (int i = 0; i < DATA.length; i++) {
            tst = StringUtil.stripXmlComments(DATA[i][0]);
            assertEquals(tst, DATA[i][1]);
        }
    }

    @Test
    public void testIsEmpty() {
        assertTrue(StringUtil.isEmpty(null));
        assertTrue(StringUtil.isEmpty(""));
        assertFalse(StringUtil.isEmpty(" "));
        assertFalse(StringUtil.isEmpty("bob"));
        assertFalse(StringUtil.isEmpty("  bob  "));
    }

    @Test
    public void testIsBlank() {
        assertTrue(StringUtil.isBlank(null));
        assertTrue(StringUtil.isBlank(""));
        assertTrue(StringUtil.isBlank(" "));
        assertFalse(StringUtil.isBlank("bob"));
        assertFalse(StringUtil.isBlank("  bob  "));
    }

    @Test
    public void testIsNotEmpty() {
        assertFalse(StringUtil.isNotEmpty(null));
        assertFalse(StringUtil.isNotEmpty(""));
        assertTrue(StringUtil.isNotEmpty(" "));
        assertTrue(StringUtil.isNotEmpty("bob"));
        assertTrue(StringUtil.isNotEmpty("  bob  "));
    }

    @Test
    public void testIsNotBlank() {
        assertFalse(StringUtil.isNotBlank(null));
        assertFalse(StringUtil.isNotBlank(""));
        assertFalse(StringUtil.isNotBlank(" "));
        assertTrue(StringUtil.isNotBlank("bob"));
        assertTrue(StringUtil.isNotBlank("  bob  "));
    }

    private static final String TEMPLATE = "StringUtilTests_template.js";

    private static final String PAUSE_TEXT = "pause.text";

    private static final String RESUME_TEXT = "resume.text";

    private static final String REFRESH_TIME = "refresh.time";

    private static final String GRAPH_IDS = "graphs";

    @Test
    public void testReplaceVariables() {
        // test using the following template..
        Map<String, String> vars = new HashMap<String, String>();
        vars.put(PAUSE_TEXT, "PAUSE");
        vars.put(REFRESH_TIME, "5");
        vars.put(RESUME_TEXT, "RESUME");
        vars.put(GRAPH_IDS, "1,2,4345");

        String tmpl = IOUtil.getResourceAsString(getClass(), TEMPLATE);
        for (Map.Entry<String, String> entry : vars.entrySet()) {
            String key = entry.getKey();
            String value = entry.getValue();
            tmpl = StringUtil.replaceVariable(tmpl, key, value);
            assertTrue(tmpl.indexOf(value) != -1);
        }
    }

    /**
     * Tests the {@link StringUtil#isWhitespace(char)} method.
     */
    @Test
    public void testWhtiespace() {
        assertTrue(StringUtil.isWhitespace(' '));
        assertTrue(StringUtil.isWhitespace('\t'));
        assertFalse(StringUtil.isWhitespace('\n'));
        assertFalse(StringUtil.isWhitespace('\r'));
        for (char i = 'a'; i < 'Z'; i++) {
            assertFalse(StringUtil.isWhitespace(i));
        }
    }

    final char TEXTQ = '"';

    final char FEILDD = ',';

    /**
     * Tests the {@link StringUtil#parseLine(String, char, char)} method.
     */
    @Test
    public void testParseLine() {
        List<Object> values;
        values = CollectionUtil.<Object> newReadOnlyList("bob", "george", 4, 23, 230948);
        parseLineTest(TEXTQ, FEILDD, values);

    }

    @Test
    public void testRandomParseLine() {
        // try random stuff..
        final Random r = new Random(17);
        final char[] replace = new char[] { TEXTQ, FEILDD };
        for (int i = 0; i < 100; i++) {
            final List<Object> values = randomList(r, 10, replace, 'a');
            parseLineTest(TEXTQ, FEILDD, values);
        }
    }

    /**
     * Tests the {@link StringUtil#parseLine(String, char, char)} methods on the
     * arguments provided.
     */
    static void parseLineTest(final char textQ, final char fieldD, final List<Object> values) {
        String csv = createCSVLine(textQ, fieldD, values);
        List<String> parsedValues = StringUtil.parseLine(csv, fieldD, textQ);
        assertEquals(parsedValues, toStringList(values));
    }

    /**
     * Create a CSV line based on the values provided.
     */
    static String createCSVLine(final char textQ, final char fieldD, final List<Object> values) {
        StringBuilder bld = new StringBuilder();
        boolean first = true;
        for (Object o : values) {
            // apply field delimiter..
            if (first) {
                first = false;
            } else {
                bld.append(fieldD);
            }
            // if its a string add the text qualifiers..
            // don't bother escape text qualifiers in the string yet..
            if (o instanceof String) {
                bld.append(textQ);
            }
            bld.append(o);
            if (o instanceof String) {
                bld.append(textQ);
            }
        }
        return bld.toString();
    }

    /**
     * Converts a {@link List} of objects to a {@link List} of {@link String}s.
     */
    static List<String> toStringList(final List<Object> list) {
        List<String> ret = new ArrayList<String>();
        for (Object o : list) {
            ret.add(o.toString());
        }
        return ret;
    }

    static List<Object> randomList(final Random r, final int size, final char[] invalid,
            final char valid) {
        List<Object> ret = new ArrayList<Object>();
        for (int i = 0; i < size; i++) {
            final Object add;
            if (r.nextBoolean()) {
                add = r.nextInt();
            } else if (r.nextBoolean()) {
                add = r.nextDouble();
            } else {
                String str = StringUtil.randomString(r, r.nextInt(30));
                for (char c : invalid) {
                    // replace all w/ 'a'..
                    str = str.replace(c, valid);
                }
                add = str;
            }
            ret.add(add);
        }
        return ret;
    }

    @Test
    public void testRandomString() {
        // just execute it because it doesn't really matter..
        String s = StringUtil.randomString();
        assertTrue(s.length() < 257);
    }

    @Test
    public void testEndsWith() {
        assertTrue(StringUtil.endsWith("afdsf", 'f'));
        assertFalse(StringUtil.endsWith(null, 'f'));
        assertFalse(StringUtil.endsWith("fadsfkj", 'f'));
    }

    private static final String PROP_TEST[] = {
      "# Some comment",
      "prop1=SomeProp",
      "prop2=OtherProp"
    };

    @Test
    public void testToProperties() {
        StringPrintWriter wrt = new StringPrintWriter();
        wrt.println(PROP_TEST);
        wrt.flush();
        String inp = wrt.getString();
        Properties prop = StringUtil.toProperties(inp);
        assertEquals(prop.get("prop1"), "SomeProp");
        assertEquals(prop.get("prop2"), "OtherProp");
    }
}

File: dotnet/FrameworkTests/StringUtilTests.cs
/*
 * ====================
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
 * 
 * Copyright 2008-2009 Sun Microsystems, Inc. All rights reserved.     
 * 
 * The contents of this file are subject to the terms of the Common Development 
 * and Distribution License("CDDL") (the "License").  You may not use this file 
 * except in compliance with the License.
 * 
 * You can obtain a copy of the License at 
 * http://opensource.org/licenses/cddl1.php
 * See the License for the specific language governing permissions and limitations 
 * under the License. 
 * 
 * When distributing the Covered Code, include this CDDL Header Notice in each file
 * and include the License file at http://opensource.org/licenses/cddl1.php.
 * If applicable, add the following below this CDDL Header, with the fields 
 * enclosed by brackets [] replaced by your own identifying information: 
 * "Portions Copyrighted [year] [name of copyright owner]"
 * ====================
 * Portions Copyrighted 2014 ForgeRock AS.
 */

using System;
using System.Collections.Generic;
using System.Text;
using Org.IdentityConnectors.Common;
using NUnit.Framework;

namespace FrameworkTests
{
    /// <summary>
    /// Description of ScriptTests.
    /// </summary>
    [TestFixture]
    public class StringUtilTests
    {
        [Test]
        public virtual void TestIndexOfDigit()
        {
            int test = 0;
            const string TEST0 = null;
            const string TEST1 = "fsadlkjffj";
            const string TEST2 = "abac2dafj";
            const string TEST3 = "fa323jf4af";

            test = StringUtil.IndexOfDigit(TEST0);
            Assert.AreEqual(test, -1);
            test = StringUtil.IndexOfDigit(TEST1);
            Assert.AreEqual(test, -1);
            test = StringUtil.IndexOfDigit(TEST2);
            Assert.AreEqual(test, 4);
            test = StringUtil.IndexOfDigit(TEST3);
            Assert.AreEqual(test, 2);
        }

        [Test]
        public virtual void TestIndexOfNonDigit()
        {
            int test = 0;
            const string TEST0 = null;
            const string TEST1 = "2131398750976";
            const string TEST2 = "21312a9320484";
            const string TEST3 = "32323aa323435";

            test = StringUtil.IndexOfNonDigit(TEST0);
            Assert.AreEqual(test, -1);
            test = StringUtil.IndexOfNonDigit(TEST1);
            Assert.AreEqual(test, -1);
            test = StringUtil.IndexOfNonDigit(TEST2);
            Assert.AreEqual(test, 5);
            test = StringUtil.IndexOfNonDigit(TEST3);
            Assert.AreEqual(test, 5);
        }

        [Test]
        public virtual void TestSubDigitString()
        {
        }

        [Test]
        public virtual void TestStripXmlAttribute()
        {
            string[][] DATA = new string[][] { new string[] { null, null, null }, new string[] { "attr='fads'", "attr", "" }, new string[] { "at1='fasd' at1=''", "at1", "" } };
            string tst = null;
            for (int i = 0; i < DATA.Length; i++)
            {
                tst = StringUtil.StripXmlAttribute(DATA[i][0], DATA[i][1]);
                Assert.AreEqual(tst, DATA[i][2]);
            }
        }

        /// <summary>
        /// Make sure it removes '\n'.
        /// </summary>
        [Test]
        public virtual void TestStripNewlines()
        {
            string[][] TESTS = new string[][] { new string[] { null, null }, new string[] { "afdslf\n", "afdslf" }, new string[] { "afds\nfadkfj", "afdsfadkfj" }, new string[] { "afds \nfadkfj", "afds fadkfj" }, new string[] { "afds\n fadkfj", "afds fadkfj" } };
            string tmp;
            foreach (string[] data in TESTS)
            {
                tmp = StringUtil.StripNewlines(data[0]);
                Assert.AreEqual(tmp, data[1]);
            }
        }

        [Test]
        public virtual void TestStripXmlComments()
        {
            string[][] DATA = new string[][] { new string[] { null, null }, new string[] { "<!--test1-->", "" }, new string[] { "test data", "test data" }, new string[] { "<!--test data", "<!--test data" }, new string[] { "test data-->", "test data-->" }, new string[] { "test data <!-- fasdkfj -->", "test data " }, new string[] { "<!-- fasdkfj --> test data", " test data" }, new string[] { "<!-- fasdkfj --> test data<!-- fadsom-->", " test data" } };

            string tst = null;
            for (int i = 0; i < DATA.Length; i++)
            {
                tst = StringUtil.StripXmlComments(DATA[i][0]);
                Assert.AreEqual(tst, DATA[i][1]);
            }
        }

        /// <summary>
        /// Tests the <seealso cref="StringUtil#parseLine(String, char, char)"/> methods on the
        /// arguments provided.
        /// </summary>
        internal static void ParseLineTest(char textQ, char fieldD, IList<object> values)
        {
            string csv = CreateCSVLine(textQ, fieldD, values);
            IList<string> parsedValues = StringUtil.ParseLine(csv, fieldD, textQ);
            Assert.AreEqual(parsedValues, ToStringList(values));
        }

        /// <summary>
        /// Create a CSV line based on the values provided.
        /// </summary>
        internal static string CreateCSVLine(char textQ, char fieldD, IList<object> values)
        {
            StringBuilder bld = new StringBuilder();
            bool first = true;
            foreach (object o in values)
            {
                // apply field delimiter..
                if (first)
                {
                    first = false;
                }
                else
                {
                    bld.Append(fieldD);
                }
                // if its a string add the text qualifiers..
                // don't bother escape text qualifiers in the string yet..
                if (o is string)
                {
                    bld.Append(textQ);
                }
                bld.Append(o);
                if (o is string)
                {
                    bld.Append(textQ);
                }
            }
            return bld.ToString();
        }

        /// <summary>
        /// Converts a <seealso cref="List"/> of objects to a <seealso cref="List"/> of <seealso cref="String"/>s.
        /// </summary>
        internal static IList<string> ToStringList(IList<object> list)
        {
            IList<string> ret = new List<string>();
            foreach (object o in list)
            {
                ret.Add(o.ToString());
            }
            return ret;
        }

        internal static IList<object> RandomList(Random r, int size, char[] invalid, char valid)
        {
            IList<object> ret = new List<object>();
            for (int i = 0; i < size; i++)
            {
                object add;
                if (r.Next(100) % 2 == 0)
                {
                    add = r.Next();
                }
                else if (r.Next(100) % 2 == 0)
                {
                    add = r.NextDouble();
                }
                else
                {
                    string str = StringUtil.RandomString(r, r.Next(30));
                    foreach (char c in invalid)
                    {
                        // replace all w/ 'a'..
                        str = str.Replace(c, valid);
                    }
                    add = str;
                }
                ret.Add(add);
            }
            return ret;
        }

        [Test]
        public virtual void TestRandomString()
        {
            // just execute it because it doesn't really matter..
            string s = StringUtil.RandomString();
            Assert.IsTrue(s.Length < 257);
        }

        [Test]
        public virtual void testEndsWith()
        {
            Assert.IsTrue(StringUtil.EndsWith("afdsf", 'f'));
            Assert.IsFalse(StringUtil.EndsWith(null, 'f'));
            Assert.IsFalse(StringUtil.EndsWith("fadsfkj", 'f'));
        }
    }
}